README.md
# gpt-assist README

GPT-Assist is a Visual Studio Code extension that integrates OpenAI's GPT-3 model into your coding workflow. It allows you to interact with GPT-3 directly from your editor, providing assistance with code writing, debugging, testing, and more.

## Features

    Selectable Prompts: Choose from a variety of prompts to guide the AI's responses, including code review, alternative solutions, code explanation, refactoring, and more.

    Project Context Awareness: GPT-Assist can read a GPT-README file that contains a description of your project, allowing it to provide more context-aware responses.

    Simulated Memory: Use keywords to refer to parts of the GPT-README file, allowing the AI to simulate memory and provide more relevant responses.

    Code Snippets: Save and insert reusable pieces of code directly from the editor.

    Documentation Generation: Automatically generate comments or documentation for your code.

    Code Formatting: Automatically format your code according to a set of style guidelines.

    Testing and Error Checking: Check your code for common errors or potential issues, write tests, and even suggest fixes.

    Integration with Version Control Systems: Manage your code versions directly from the editor.

### Planned Implementation Steps

    Step 1: Implement selectable prompts. This will involve creating a dropdown menu or command palette where the user can choose the type of prompt they want to use.

    Step 2: Add the ability to read from a GPT-README file. This will involve parsing the file and incorporating its contents into the AI's responses.

    Step 3: Implement simulated memory. This will involve recognizing and substituting keywords in the user's input, and allowing the AI to generate its own prompts.

    Step 4: Add code snippet functionality. This will involve creating a way for the user to save and insert reusable pieces of code.

    Step 5: Implement documentation generation. This will involve using the AI to automatically generate comments or documentation based on the user's code.

    Step 6: Add code formatting functionality. This will involve integrating an existing code formatter into the extension.

    Step 7: Implement testing and error checking. This will involve using the AI to check the user's code for errors, write tests, and suggest fixes.

    Step 8: Add integration with version control systems. This will involve adding features for working with systems like Git.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Working with Markdown

You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "gpt-assist" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

test/suite/database.test.js
const assert = require('assert');
const { insertMemory, calculateRecencyScores, retrieveMemory, insertFunction, retrieveFunction } = require('../database');

suite('Database Test Suite', () => {
    suite('Memories', () => {
        test('should insert a memory into the database', async () => {
            await insertMemory('prompt', 'output', 'summary', 'importance');
            let memory = await retrieveMemory('prompt');
            assert(memory);
            assert.strictEqual(memory.prompt, 'prompt');
        });

        test('should calculate recency scores', async () => {
            await insertMemory('prompt1', 'output1', 'summary1', 'importance1');
            await insertMemory('prompt2', 'output2', 'summary2', 'importance2');
            await calculateRecencyScores();
            let memory = await retrieveMemory('prompt1');
            assert(memory);
            assert(memory.recencyScore);
        });
    });

    suite('Functions', () => {
        test('should insert a function into the database', async () => {
            let embedding = [0.5, 0.4, 0.6];
            await insertFunction('path', 'name', embedding);
            let func = await retrieveFunction('name');
            assert(func);
            assert.strictEqual(func.name, 'name');
            assert.deepStrictEqual(func.embedding, embedding);
        });
    });
});


database.js
const Datastore = require('nedb-promises');
const path = require('path');
const vscode = require('vscode');
const fs = require('fs');

let dbPath = path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.gpt-assist.db');
let db = fs.existsSync(dbPath) ? Datastore.create({ filename: dbPath }) : null;

let funcDbPath = path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.gpt-assist-functions.db');
let funcDb = fs.existsSync(funcDbPath) ? Datastore.create({ filename: funcDbPath }) : null;

let metaDbPath = path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.gpt-assist-meta.db');
let metaDb = fs.existsSync(metaDbPath) ? Datastore.create({ filename: metaDbPath }) : null;

/**
 * Sets the initialization flag for a file in the database.
 * @param {string} filePath - The path to the file.
 * @param {boolean} isInitialized - Whether or not the file has been indexed.
 */
async function setFileInitializationFlag(filePath, isInitialized) {
    try {
        if (!metaDb) {
            metaDb = Datastore.create({ filename: metaDbPath });
        }

        // Check if the flag already exists
        let flag = await metaDb.findOne({ key: filePath });
        
        if (flag) {
            // If the flag exists, update it
            await metaDb.update({ key: filePath }, { $set: { value: isInitialized } });
        } else {
            // If the flag does not exist, insert it
            await metaDb.insert({ key: filePath, value: isInitialized });
        }
    } catch (err) {
        console.error('Error setting file initialization flag:', err);
    }
}

/**
 * Gets the initialization flag for a file from the database.
 * @param {string} filePath - The path to the file.
 * @returns {boolean} Whether or not the file has been indexed.
 */
async function getFileInitializationFlag(filePath) {
    try {
        if (!metaDb) {
            return null;
        }

        // Retrieve the flag
        let flag = await metaDb.findOne({ key: filePath });
        return flag ? flag.value : null;
    } catch (err) {
        console.error('Error getting file initialization flag:', err);
    }
}

/**
 * Inserts a new function into the database.
 * @param {string} path - The file path.
 * @param {string} name - The function name.
 * @param {Array} embedding - The embedding.
 */
async function insertFunction(path, name, embedding) {
    try {
        if (!funcDb) {
            funcDb = Datastore.create({ filename: funcDbPath });
        }

        let newFunction = { path: path, name: name, embedding: embedding.embedding, lastModified: new Date() };
        await funcDb.insert(newFunction);
    } catch (err) {
        console.error('Error inserting function:', err);
    }
}

/**
 * Retrieves a function from the database based on its name.
 * @param {string} name - The function name.
 * @returns {Object} The function.
 */
async function retrieveFunction(name) {
    try {
        if (!funcDb) {
            return null;
        }

        // Retrieve the function
        let func = await funcDb.findOne({ name: name });
        return func;
    } catch (err) {
        console.error('Error retrieving function:', err);
    }
}

/**
 * Inserts a new memory into the database.
 * @param {string} prompt - The prompt text.
 * @param {string} output - The output text.
 */
async function insertMemory(prompt, output, summary, importance) {
    try {
        if (!db) {
            db = Datastore.create({ filename: dbPath });
        }
        let now = Date.now();
        let lastAccessedMemory = await db.findOne({}).sort({ accessedAt: -1 });
        let timeDifference = now - (lastAccessedMemory ? lastAccessedMemory.accessedAt : now);
        let newMemory = { prompt: prompt, output: output, summary: summary, importance: importance, accessedAt: now, timeDifference: timeDifference, recencyScore: 1 };
        await db.insert(newMemory);
    } catch (err) {
        console.error('Error inserting memory:', err);
    }
}

/**
 * Calculates the recency scores for all memories in the database.
 */
async function calculateRecencyScores() {
    try {
        if (!db) {
            return;
        }
        let now = Date.now();
        let memories = await db.find({}).sort({ accessedAt: -1 }).limit(2);
        let cumulativeTimeDifference = memories.length > 1 ? now - memories[1].accessedAt : 0;
        memories = await db.find({}).sort({ accessedAt: 1 });
        for (let memory of memories) {
            cumulativeTimeDifference += memory.timeDifference;
            let recencyScore = Math.pow(0.99, cumulativeTimeDifference / (1000 * 60 * 60)); // Exponential decay over hours
            await db.update({ _id: memory._id }, { $set: { recencyScore: recencyScore } });
        }
    } catch (err) {
        console.error('Error calculating recency scores:', err);
    }
}

/**
 * Retrieves the most recent memory from the database.
 * @param {string} query - The query text.
 * @returns {Object} The most recent memory.
 */
async function retrieveMemory(query) {
    try {
        if (!db) {
            return null;
        }
        // Retrieve the most recent memory
        let memory = await db.findOne({}).sort({ accessedAt: -1 });
        if (!memory) {
            return null;
        }
        // Update the accessedAt and timeDifference properties of the memory
        let now = Date.now();
        await db.update({ _id: memory._id }, { $set: { accessedAt: now } });
        let nextMemory = await db.findOne({}).sort({ accessedAt: -1 }).skip(1);
        if (nextMemory) {
            let timeDifference = now - nextMemory.accessedAt;
            await db.update({ _id: nextMemory._id }, { $set: { timeDifference: timeDifference } });
        }
        // Return the memory
        return memory;
    } catch (err) {
        console.error('Error retrieving memory:', err);
    }
}

async function insertMemoryFromFile() {
    const filePath = path.join(__dirname, 'dummy_prompt.txt');
    const fileContents = fs.readFileSync(filePath, 'utf8');
    const [prompt, output] = fileContents.split('\n');
    let summaryMatch = output.match(/Summary: (.*?)\n/);
    let importanceMatch = output.match(/Importance: (.*?)\n/);
    let summary = summaryMatch ? summaryMatch[1] : '';
    let importance = importanceMatch ? importanceMatch[1] : '';
    insertMemory(prompt, output, summary, importance);
  }
  
  
module.exports = { insertMemory, calculateRecencyScores, retrieveMemory, insertMemoryFromFile,
    insertFunction, retrieveFunction,
    setFileInitializationFlag, getFileInitializationFlag };


embedder.js
const { spawn } = require('child_process');

async function generateEmbedding(text) {
    return new Promise((resolve, reject) => {
        const python = spawn('python', ['./embedder.py']);
        let dataString = '';

        python.stdout.on('data', (data) => {
            dataString += data.toString();
        });

        python.stderr.on('data', (data) => {
            console.error(`Python error: ${data}`);
        });

        python.stdout.on('end', () => {
            try {
                resolve(JSON.parse(dataString));
            } catch (error) {
                reject(error);
            }
        });

        python.stdin.write(text);
        python.stdin.end();
    });
}
if (require.main === module) {
    // If the script is run directly, perform a test
    generateEmbedding(`async function generateEmbedding(text) {
        return new Promise((resolve, reject) => {
            const python = spawn('python', ['./embedder.py']);
            let dataString = '';
    
            python.stdout.on('data', (data) => {
                dataString += data.toString();
            });
    
            python.stdout.on('end', () => {
                try {
                    resolve(JSON.parse(dataString));
                } catch (error) {
                    reject(error);
                }
            });
    
            python.stdin.write(text);
            python.stdin.end();
        });
    }`)
        .then(embedding => console.log(embedding))
        .catch(error => console.error(error));
}

module.exports = { generateEmbedding };


embedder.py
import tensorflow_hub as hub
import tensorflow as tf
import numpy as np
import sys
import json
import os

print('{"test": 123}')
sys.stdout.flush()

# Define the URL of the model and the local directory to cache the model
MODEL_URL = "https://tfhub.dev/google/universal-sentence-encoder/4"
MODEL_DIR = './embedding'
if not os.path.exists(MODEL_DIR):
    os.makedirs(MODEL_DIR)
    os.environ['TFHUB_CACHE_DIR'] = MODEL_DIR

# Load the Universal Sentence Encoder's TF Hub module
model = hub.load(MODEL_URL)

def embed_text(text):
    # Compute a representation for the provided texts
    embedding = model([text])

    # We return the embedding as a numpy array, which we convert to a list so it can be JSON serialized
    return np.array(embedding).tolist()

if __name__ == "__main__":
    # Accept input from standard input
    with open('debug.txt', 'w') as f:
        f.write('About to read from stdin...\n')
        text = sys.stdin.read()
        f.write(f'Read from stdin: {text}\n')

    # Get the embedding
    embedding = embed_text(text)

    # Create a JSON object with the embedding
    result = {"embedding": embedding}

    # Print the JSON object
    print(json.dumps(result))

extension.js
// Importing required modules
const vscode = require('vscode');
const tasks = require('./tasks.json');
const { callOpenAI } = require('./openai');
const { getTotalTokens, tokenStatusBarItem } = require('./vars');
const { insertMemoryFromFile, getFileInitializationFlag, setFileInitializationFlag, insertFunction } = require('./database');
const { parseWorkspace } = require('./codeParser');
const { generateEmbedding } = require('./embedder');

/**
 * Function to activate the extension
 * @param {vscode.ExtensionContext} context - The context in which the extension is run
 */
function activate(context) {
    console.log('Congratulations, your extension "gpt-assist" is now active!');

    tokenStatusBarItem.text = `Tokens used: ${getTotalTokens()}`;
    tokenStatusBarItem.show();
    context.subscriptions.push(tokenStatusBarItem);

    let helloWorldDisposable = vscode.commands.registerCommand('gpt-assist.helloWorld', function () {
        vscode.window.showInformationMessage('Hello World from gpt-assist!');
    });

    let openAIDisposable = vscode.commands.registerCommand('gpt-assist.call_gpt', async function () {
        const editor = vscode.window.activeTextEditor;

        if (editor) {
            let document = editor.document;
            let selection = editor.selection;

            // Get the word within the selection
            let text = document.getText(selection);

            // Ask the user to choose a task
            const taskOptions = Object.keys(tasks);
            taskOptions.push('Custom Prompt');
            let task = await vscode.window.showQuickPick(taskOptions, { placeHolder: 'Choose a task' });
            if (task === 'Custom Prompt') {
                task = await vscode.window.showInputBox({ prompt: 'Enter your custom prompt' });
            }

            else {
                task = tasks[task]
            }

            try {
                const response = await callOpenAI(text, task, true, true);
                // Display the OpenAI response to the user
                vscode.window.showInformationMessage(response);
            } catch (err) {
                vscode.window.showErrorMessage('Error: ' + err);
            }
        }
    });

    let insertMemoryCommand = vscode.commands.registerCommand('gpt-assist.insertMemory', async () => {
        try {
          await insertMemoryFromFile();
          vscode.window.showInformationMessage('Memory inserted successfully.');
        } catch (error) {
          vscode.window.showErrorMessage('Failed to insert memory: ' + error.message);
        }
    });

    let initializeWorkspaceDisposable = vscode.commands.registerCommand('gpt-assist.initializeWorkspace', async function () {
        const workspaceFolder = vscode.workspace.workspaceFolders[0].uri.fsPath;
    
        await parseWorkspace(workspaceFolder, async (filePath, funcName, funcCode) => {
            let isInitialized = await getFileInitializationFlag(filePath);
            
            if (!isInitialized) {
                // generate embedding for function
                let embedding = await generateEmbedding(funcCode);
    
                // insert the function into your database
                await insertFunction(filePath, funcName, embedding);
    
                // Mark the file as initialized
                await setFileInitializationFlag(filePath, true);
    
                vscode.window.showInformationMessage(`File ${filePath} has been initialized.`);
            } else {
                vscode.window.showInformationMessage(`File ${filePath} has already been initialized.`);
            }
        });
    
        vscode.window.showInformationMessage('Workspace initialization complete.');
    });

    context.subscriptions.push(helloWorldDisposable, openAIDisposable, insertMemoryCommand, initializeWorkspaceDisposable);
}

/**
 * Function to deactivate the extension
 */
function deactivate() {}

module.exports = {
	activate,
	deactivate
}


codeParser.js
// Import required modules
const esprima = require('esprima');
const estraverse = require('estraverse');
const fs = require('fs');
const path = require('path');

/**
 * Parses a JavaScript file and extracts information about function declarations.
 * @param {string} filePath - The path of the JavaScript file to parse.
 * @returns {Array} - An array of objects containing function names and their corresponding code.
 */
async function parseFile(filePath) {
    let code = fs.readFileSync(filePath, 'utf8');
    let ast = esprima.parse(code, { range: true });
    let functions = [];

    estraverse.traverse(ast, {
        enter: function (node) {
            if (node.type === 'FunctionDeclaration') {
                let funcCode = code.substring(node.range[0], node.range[1]);
                let funcName = node.id ? node.id.name : null;
                functions.push({ name: funcName, code: funcCode });
            }
        }
    });

    return functions;
}

/**
 * Parses all JavaScript files in a workspace directory and invokes a callback function for each function found.
 * @param {string} workspacePath - The path of the workspace directory.
 * @param {Function} callback - The callback function to invoke for each function found.
 */
async function parseWorkspace(workspacePath, callback) {
    let files = fs.readdirSync(workspacePath);

    for (let file of files) {
        if (path.extname(file) !== '.js') continue;

        let filePath = path.join(workspacePath, file);
        let functions = await parseFile(filePath);

        for (let func of functions) {
            callback(filePath, func.name, func.code);
        }
    }
}

// Export the parseFile and parseWorkspace functions
module.exports = { parseFile, parseWorkspace };


openai.js
const { Configuration, OpenAIApi } = require('openai');
const { get_encoding } = require('@dqbd/tiktoken');
const vscode = require('vscode');
const fs = require('fs');
const path = require('path');
const { setTotalTokens, getTotalTokens, tokenStatusBarItem, projectDescription } = require('./vars');
const { insertMemory, retrieveMemory } = require('./database');

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);
let model = "dummy" // vscode.workspace.getConfiguration('gpt-assist').get('model');

/**
 * Function to call OpenAI API and get the response
 * @param {string} text - The text selected by the user to be sent to the OpenAI API
 * @param {string} task - The prompt associated to the task to be performed by the OpenAI API
 * @returns {string} output - The response from the OpenAI API
 */
async function callOpenAI(text, task, retrieveMemories, insertNewMemory) {
  let memory;
  let prompt = task;
  const outputChannel = vscode.window.createOutputChannel('GPT-Assist');

  // Retrieve a relevant memory from the database if required
  if (retrieveMemories) {
    memory = await retrieveMemory({ text: text , task: task});
    prompt += memory ? '\nSummary of previous interaction: ' + memory.summary + '\n' : '';
  }

  prompt += text;

   // Add a short prompt asking the model to rate the importance of the interaction
   if (insertNewMemory) {
    prompt += '\n\nPlease provide a short summary of this interaction and rate its importance to the project described below:\n' + projectDescription;
  }

  // Get the tokenizer and calculate the token count only when the model is not "dummy"
  let tokenCount;
  let cost;
  if (model !== "dummy") {
    const enc = get_encoding(model);
    tokenCount = enc.encode(prompt).length;
    cost = (tokenCount / 1000) * 0.01; // 1 cent per 1000 tokens
    enc.free();
  }

  let output;
  if (model !== "dummy") {
    // Inform the user about the number of tokens and the cost
    const userConfirmation = await vscode.window.showInformationMessage(
      `Your prompt contains ${tokenCount} tokens. The cost will be $${cost.toFixed(2)}. Do you want to proceed?`,
      { modal: true },
      'Yes'
    );

    // If the user confirms, call the OpenAI API
    if (userConfirmation === 'Yes') {
      const response = await openai.createCompletion({
        model: model,
        prompt: prompt,
        max_tokens: 60,
      });
      output = response.data.choices[0].text;
    } else {
      return; // Exit the function if the user doesn't confirm
    }
  } else {
    // Display the prompt in an output channel
    outputChannel.clear();
    const filePath = path.join(__dirname, 'dummy_prompt.txt');
    fs.writeFileSync(filePath, prompt);
    output = 'The prompt has been written to dummy_prompt.txt. Please provide an output.';
    outputChannel.appendLine(output);
    outputChannel.show();
    output = undefined
    }

  // Store the prompt and output as a new memory in the database if required
  if (insertNewMemory && output !== undefined) {
    // Parse the summary and importance from the output
    let summaryMatch = output.match(/Summary: (.*?)\n/);
    let importanceMatch = output.match(/Importance: (.*?)\n/);
    let summary = summaryMatch ? summaryMatch[1] : '';
    let importance = importanceMatch ? importanceMatch[1] : '';
    insertMemory(prompt, output, summary, importance);
  }

  setTotalTokens(getTotalTokens() + tokenCount);
  tokenStatusBarItem.text = `Tokens used: ${getTotalTokens()}`;

  // Check if the markdown file exists, if not, create it
  const filePath = path.join(__dirname, 'conversation.md');
  if (!fs.existsSync(filePath)) {
    fs.writeFileSync(filePath, '');
  }

  // Append the input and output to the markdown file
  fs.appendFileSync(
    filePath,
    `---\n\n**User:**\n${task}:\n${text}\n\n_AI:_\n${output}\n\n`
  );

  // Open the markdown file in a new editor
  const document = await vscode.workspace.openTextDocument(filePath);
  await vscode.window.showTextDocument(document, vscode.ViewColumn.Beside);

  return output;
}

module.exports = { callOpenAI };
